File: snippets/robot-nervous-system/blackboard/cxx/src/reader.cpp
#include "distance.hpp"
#include "iox2/bb/static_string.hpp"
#include "iox2/iceoryx2.hpp"

#include <iostream>
#include <utility>

namespace {
// stands in for the platform-specific ultrasonic sensor read
auto get_ultra_sonic_sensor_distance() -> double {
    return 0.42; // NOLINT
}
} // namespace

auto main() -> int {
    using namespace iox2;

    auto node = NodeBuilder().create<ServiceType::Ipc>().value();

    // the distance publisher is reused from the publish-subscribe example
    auto publisher = node.service_builder(ServiceName::create("distance_to_obstacle").value())
                         .publish_subscribe<Distance>()
                         .open_or_create()
                         .value()
                         .publisher_builder()
                         .create()
                         .value();

    // snippet:start service
    using KeyType = bb::StaticString<50>;
    auto service =
        node.service_builder(ServiceName::create("global_config").value()).blackboard_opener<KeyType>().open().value();
    // snippet:end service

    // snippet:start reader
    auto reader = service.reader_builder().create().value();
    // snippet:end reader

    // snippet:start handle
    auto us_sensor_key = bb::StaticString<50>::from_utf8("ultra_sonic_sensor_update_rate_in_ms").value();
    auto update_rate_handle = reader.template entry<uint32_t>(us_sensor_key).value();
    // snippet:end handle

    // snippet:start sensor-loop
    while (node.wait(iox2::bb::Duration::from_millis(*update_rate_handle.get())).has_value()) {
        auto sample = publisher.loan_uninit().value();

        auto initialized_sample = sample.write_payload(Distance { get_ultra_sonic_sensor_distance(), 42.0 }); // NOLINT

        send(std::move(initialized_sample)).value();
    }
    // snippet:end sensor-loop

    return 0;
}

File: snippets/robot-nervous-system/blackboard/rust/src/bin/reader.rs
use blackboard::distance::Distance;
use core::time::Duration;
use iceoryx2_bb_container::string::StaticString;

// stands in for the platform-specific ultrasonic sensor read
fn get_ultra_sonic_sensor_distance() -> f64 {
    0.42
}

fn main() -> Result<(), Box<dyn core::error::Error>> {
    use iceoryx2::prelude::*;

    let node = NodeBuilder::new().create::<ipc::Service>()?;

    // the distance publisher is reused from the publish-subscribe example
    let publisher = node
        .service_builder(&"distance_to_obstacle".try_into()?)
        .publish_subscribe::<Distance>()
        .open_or_create()?
        .publisher_builder()
        .create()?;

    // snippet:start service
    type KeyType = StaticString<50>;
    let service = node
        .service_builder(&"global_config".try_into()?)
        .blackboard_opener::<KeyType>()
        .open()?;
    // snippet:end service

    // snippet:start reader
    let reader = service.reader_builder().create()?;
    // snippet:end reader

    // snippet:start handle
    let update_rate_handle =
        reader.entry::<u32>(&"ultra_sonic_sensor_update_rate_in_ms".try_into()?)?;
    // snippet:end handle

    // snippet:start sensor-loop
    while node
        .wait(Duration::from_millis(*update_rate_handle.get() as u64))
        .is_ok()
    {
        let sample = publisher.loan_uninit()?;

        let sample = sample.write_payload(Distance {
            distance_in_meters: get_ultra_sonic_sensor_distance(),
            some_other_property: 42.0,
        });

        sample.send()?;
    }
    // snippet:end sensor-loop

    Ok(())
}
